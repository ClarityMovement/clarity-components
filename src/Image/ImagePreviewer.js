import React from 'react';
import T from 'prop-types';
import Image from 'react-image';
import LightBox from 'react-image-lightbox';
import Spinner from 'react-md-spinner';
import { compose, withState, withHandlers } from 'recompose';
import { omitProps, embedHandler } from 'react-render-counter/hocs';

const styles = {
  root: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#f7f7f7',
  },
  image: { width: '100%', height: '100%' },
};

const ImagePreviewerRenderer = ({
  src,
  style,
  modalZIndex,
  open,
  openLightBox,
  closeLightBox,
  ...rest
}) => (
  <div style={{ ...styles.root, ...style }}>
    <Image
      {...rest}
      style={styles.image}
      src={src}
      onClick={openLightBox}
      loader={<Spinner />}
    />
    {open ? (
      <LightBox
        mainSrc={src}
        onCloseRequest={closeLightBox}
        reactModalStyle={{
          overlay: { zIndex: modalZIndex },
        }}
      />
    ) : null }
  </div>
);

ImagePreviewerRenderer.propTypes = {
  closeLightBox: T.func,
  open: T.bool,
  openLightBox: T.func,
  src: T.string.isRequired,
  style: T.objectOf(T.any),
  modalZIndex: T.number,
};

ImagePreviewerRenderer.defaultProps = {
  closeLightBox: undefined,
  open: false,
  openLightBox: undefined,
  style: styles.image,
  modalZIndex: 9999,
};

export const handlers = {
  openLightBox: ({ onSetOpen }) => () => onSetOpen(true),
  closeLightBox: ({ onSetOpen }) => () => onSetOpen(false),
};

const withImagePreviewer = compose(
  withState('open', 'setOpen', ({ initialOpen }) => initialOpen),
  embedHandler('setOpen', 'onSetOpen'),
  withHandlers(handlers),
  omitProps(['setOpen', 'onSetOpen', 'initialOpen']),
);

export {
  ImagePreviewerRenderer,
  withImagePreviewer,
};

const ImagePreviewer = withImagePreviewer(ImagePreviewerRenderer);
ImagePreviewer.displayName = 'ImagePreviewer';

export default ImagePreviewer;
