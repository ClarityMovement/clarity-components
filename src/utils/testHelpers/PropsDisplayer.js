import React, { Component } from 'react';

// This Component can't render deeply, must work with shallow renderer.
// Recommend to use shallowWithUntil
const PropsDisplayer = props => (
  <div name="PropsDisplayer">
    <Component {...props} />
  </div>
);

export default PropsDisplayer;
