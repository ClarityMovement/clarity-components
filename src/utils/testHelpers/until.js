import { shallow } from 'enzyme';

// Copied from gist
// https://gist.github.com/matthieuprat/5fd37abbd4a4002e6cfe0c73ae54cda8
export default function until(selector, options = this.options) {
  if (this.isEmptyRender() || typeof this.node.type === 'string') { return this; }

  let context = options.context;

  const instance = this.instance();
  if (instance.getChildContext) {
    context = {
      ...context,
      ...instance.getChildContext(),
    };
  }

  const wrapper = this.shallow({ context });

  if (!this.is(selector)) {
    return until.call(wrapper, selector, { context });
  }

  return wrapper;
}

export const shallowWithUntil = (...args) => {
  const wrapper = shallow(...args);
  return Object.assign(wrapper, { until });
};
