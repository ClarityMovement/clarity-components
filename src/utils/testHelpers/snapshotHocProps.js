import React from 'react';

import enzymeToJson from 'enzyme-to-json';
import { shallowWithUntil as shallow } from './until';

const snapshotPropsMapperHoc = (hoc, inputProps) => {
  // This Component can't render deeply, must work with shallow renderer.
  const PropsDisplayer = outputProps => (
    <div name="props-displayer">
      <input-props {...inputProps} />
      <output-props {...outputProps} />
    </div>
  );
  const NewComponent = hoc(PropsDisplayer);
  const wrapper = shallow(<NewComponent {...inputProps} />).until(PropsDisplayer);
  expect(enzymeToJson(wrapper)).toMatchSnapshot();
};

export default snapshotPropsMapperHoc;
