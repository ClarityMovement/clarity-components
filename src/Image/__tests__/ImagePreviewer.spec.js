/* eslint-env jest */
import React from 'react';
import enzymeToJson from 'enzyme-to-json';
import { shallowWithUntil as shallow } from '../../utils/testHelpers/until';
import ImagePreviewer, { ImagePreviewerRenderer } from '../ImagePreviewer';

const defaultProps = {
  ...ImagePreviewerRenderer.defaultProps,
  src: 'http://pic.example.com/200/200',
  width: '90%',
};

describe('<ImagePreviewer {...props} />: elements tree', () => {
  describe('structures', () => {
    let props;

    afterEach(() => {
      const wrapper = shallow(<ImagePreviewer {...props} />).until(ImagePreviewerRenderer);
      expect(enzymeToJson(wrapper)).toMatchSnapshot();
    });

    test('with default props', () => {
      props = defaultProps;
    });

    test('with initialOpen controlled', () => {
      props = {
        ...defaultProps,
        initialOpen: true,
      };
    });
  });

  describe('interactions', () => {
    let root;
    const event = {
      stopPropagation: jest.fn(),
    };
    const getWrapperProps = () => root.until(ImagePreviewerRenderer).instance().props;

    beforeEach(() => event.stopPropagation.mockClear());

    describe('with default props', () => {
      test('openLightBox && closeLightBox', () => {
        root = shallow(<ImagePreviewer {...defaultProps} />);
        const { openLightBox, closeLightBox } = getWrapperProps();

        expect(getWrapperProps().open).toBe(false);
        openLightBox(event);
        expect(event.stopPropagation).toHaveBeenCalledTimes(1);
        expect(getWrapperProps().open).toBe(true);
        closeLightBox(event);
        expect(event.stopPropagation).toHaveBeenCalledTimes(2);
        expect(getWrapperProps().open).toBe(false);
      });
    });

    describe('with onSetOpen controlled', () => {
      test('openLightBox && closeLightBox', () => {
        const onSetOpen = jest.fn();
        root = shallow(<ImagePreviewer {...defaultProps} onSetOpen={onSetOpen} />);
        const { openLightBox, closeLightBox } = getWrapperProps();

        openLightBox(event);
        expect(event.stopPropagation).toHaveBeenCalledTimes(1);
        expect(onSetOpen).toHaveBeenCalledTimes(1);
        expect(onSetOpen).toHaveBeenCalledWith(true);

        closeLightBox(event);
        expect(event.stopPropagation).toHaveBeenCalledTimes(2);
        expect(onSetOpen).toHaveBeenCalledTimes(2);
        expect(onSetOpen).toHaveBeenCalledWith(false);
      });
    });
  });
});
