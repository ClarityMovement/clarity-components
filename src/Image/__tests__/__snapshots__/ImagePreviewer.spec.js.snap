// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagePreviewer {...props} />: elements tree structures with default props 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f7f7f7",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Img
    loader={
      <MDSpinner
        color1="rgb(66, 165, 245)"
        color2="rgb(239, 83, 80)"
        color3="rgb(253, 216, 53)"
        color4="rgb(76, 175, 80)"
        duration={1333}
        size={28}
      />
    }
    onTouchTap={[Function]}
    src="http://pic.example.com/200/200"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
    unloader={false}
    width="90%"
  />
</div>
`;

exports[`<ImagePreviewer {...props} />: elements tree structures with initialOpen controlled 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f7f7f7",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Img
    loader={
      <MDSpinner
        color1="rgb(66, 165, 245)"
        color2="rgb(239, 83, 80)"
        color3="rgb(253, 216, 53)"
        color4="rgb(76, 175, 80)"
        duration={1333}
        size={28}
      />
    }
    onTouchTap={[Function]}
    src="http://pic.example.com/200/200"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
    unloader={false}
    width="90%"
  />
  <ReactImageLightbox
    animationDisabled={false}
    animationDuration={300}
    animationOnKeyInput={false}
    clickOutsideToClose={true}
    closeLabel="Close lightbox"
    discourageDownloads={false}
    enableZoom={true}
    imagePadding={10}
    keyRepeatKeyupBonus={40}
    keyRepeatLimit={180}
    mainSrc="http://pic.example.com/200/200"
    nextLabel="Next image"
    onAfterOpen={[Function]}
    onCloseRequest={[Function]}
    onImageLoadError={[Function]}
    onMoveNextRequest={[Function]}
    onMovePrevRequest={[Function]}
    prevLabel="Previous image"
    reactModalStyle={
      Object {
        "overlay": Object {
          "zIndex": 9999,
        },
      }
    }
    wrapperClassName=""
    zoomInLabel="Zoom in"
    zoomOutLabel="Zoom out"
  />
</div>
`;
